Mathlib/Tactic/PushNeg.lean

```diff
@@ -0,0 +1,137 @@
+      | false => return mkSimpStep (.forallE `_ p (mkNot q) default) (←mkAppM ``not_and_eq #[p, q])
+      | true  => return mkSimpStep (mkOr (mkNot p) (mkNot q)) (←mkAppM ``not_and_distrib_eq #[p, q])
+  | (``Or, #[p, q]) =>
+      return mkSimpStep (mkAnd (mkNot p) (mkNot q)) (←mkAppM ``not_or_eq #[p, q])
+  | (``Eq, #[_ty, e₁, e₂]) =>
+      return Simp.Step.visit { expr := ← mkAppM ``Ne #[e₁, e₂] }
+  | (``Ne, #[_ty, e₁, e₂]) =>
+      return mkSimpStep (← mkAppM ``Eq #[e₁, e₂]) (← mkAppM ``not_ne_eq #[e₁, e₂])
+  | (``LE.le, #[_ty, _inst, e₁, e₂]) =>
+      return mkSimpStep (← mkAppM ``LT.lt #[e₂, e₁]) (← mkAppM ``not_le_eq #[e₁, e₂])
+  | (``LT.lt, #[_ty, _inst, e₁, e₂]) =>
+      return mkSimpStep (← mkAppM ``LE.le #[e₂, e₁]) (← mkAppM ``not_lt_eq #[e₁, e₂])
+  | (``Exists, #[_, .lam n typ bo bi]) =>
+      return mkSimpStep (.forallE n typ (mkNot bo) bi)
+                        (← mkAppM ``not_exists_eq #[.lam n typ bo bi])
+  | (``Exists, #[_, _]) =>
+      return none
+  | _ => match ex with
+         | .forallE name ty body binfo => do
+           let body' : Expr := .lam name ty (mkNot body) binfo
+           let body'' : Expr := .lam name ty body binfo
+           return mkSimpStep (← mkAppM ``Exists #[body']) (← mkAppM ``not_forall_eq #[body''])
+         | _ => return none
+
+/-- Recursively push negations at the top level of the current expression. This is needed
+to handle e.g. triple negation. -/
+partial def transformNegation (e : Expr) : SimpM Simp.Step := do
+  let Simp.Step.visit r₁ ← transformNegationStep e | return Simp.Step.visit { expr := e }
+  match r₁.proof? with
+  | none => return Simp.Step.visit r₁
+  | some _ => do
+      let Simp.Step.visit r₂ ← transformNegation r₁.expr | return Simp.Step.visit r₁
+      return Simp.Step.visit (← Simp.mkEqTrans r₁ r₂)
+
+/-- Execute main loop of `push_neg` at the main goal. -/
+def pushNegTarget : TacticM Unit := withMainContext do
+  let myctx : Simp.Context :=
+    { config := { eta := true },
+      simpTheorems := #[ ]
+      congrTheorems := (← getSimpCongrTheorems) }
+  let goal ← getMainGoal
+  let tgt ← instantiateMVars (← getMVarType goal)
+  let myres := ← Simp.main tgt myctx (methods := { pre := transformNegation })
+  replaceMainGoal [← applySimpResultToTarget goal tgt myres]
+
+/-- Execute main loop of `push_neg` at a local hypothesis. -/
+def pushNegLocalDecl (fvarId : FVarId): TacticM Unit := withMainContext do
+  let ldecl ← getLocalDecl fvarId
+  if ldecl.isAuxDecl then return
+  let tgt := (← getLocalDecl fvarId).type
+  let myctx : Simp.Context :=
+    { config := { eta := true },
+      simpTheorems := #[ ]
+      congrTheorems := (← getSimpCongrTheorems) }
+  let goal ← getMainGoal
+  let myres := ← Simp.main tgt myctx (methods := { pre := transformNegation })
+  let some ⟨_, newGoal⟩ ← applySimpResultToLocalDecl goal fvarId myres False | throwError "fail"
+  replaceMainGoal [newGoal]
+
+/--
+Push negations in the goal of some assumption.
```

Comment:
```suggestion
Push negations into the conclusion of a hypothesis.
```
I think using "goal" here is quite confusing. Also, it's probably more readable if we are consistent about hypothesis/assumption.

Reply:
LGTM otherwise!